File: Algorithms/Strings/Bear and Steady Gene.py
import os

#
# Complete the 'steadyGene' function below.
#
# The function is expected to return an INTEGER.
# The function accepts STRING gene as parameter.
#

def steadyGene(gene):
    # Write your code here
    from collections import Counter

    n = len(gene)
    ideal_count = n // 4
    count = Counter(gene)

    excess = {char: max(0, count[char] - ideal_count) for char in "ACGT"}

    if all(v == 0 for v in excess.values()):
        return 0

    min_length = (n)

    left = 0

    for right in range(n):
        count[gene[right]] -= 1

        while all(count[char] <= ideal_count for char in "ACGT"):
            min_length = min(min_length, right - left + 1)
            count[gene[left]] += 1
            left += 1

    return min_length


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input().strip())
    gene = input()

    result = steadyGene(gene)

    fptr.write(str(result) + '\n')
    fptr.close()

File: Algorithms/Strings/Bear and Steady Gene.php
<?php

/*
 * Complete the 'steadyGene' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts STRING gene as parameter.
 */

function steadyGene($gene) {
    # Write your code here
    $n = strlen($gene);
    $ideal_count = $n / 4;
    $count = array_count_values(str_split($gene));

    foreach (['A', 'C', 'G', 'T'] as $char) {
        if (!isset($count[$char])) {
            $count[$char] = 0;
        }
    }

    $excess = array('A' => 0, 'C' => 0, 'G' => 0, 'T' => 0);
    foreach ($count as $char => $c) {
        if ($c > $ideal_count) {
            $excess[$char] = $c - $ideal_count;
        }
    }

    if (array_sum($excess) == 0) {
        return 0;
    }

    $min_length = $n;
    $left = 0;

    for ($right = 0; $right < $n; $right++) {
        $count[$gene[$right]]--;

        while (allBelowOrEqual($count, $ideal_count)) {
            $min_length = min($min_length, $right - $left + 1);
            $count[$gene[$left]]++;
            $left++;
        }
    }

    return $min_length;
}

function allBelowOrEqual($count, $limit) {
    foreach (['A', 'C', 'G', 'T'] as $char) {
        if ($count[$char] > $limit) {
            return False;
        }
    }
    return True;
}

$fptr = fopen(getenv("OUTPUT_PATH"), "w");

$n = intval(trim(fgets(STDIN)));

$gene = rtrim(fgets(STDIN), "\r\n");

$result = steadyGene($gene);

fwrite($fptr, $result . "\n");
fclose($fptr);
